runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,339 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+      {
+      traceMsg(cg->comp(), "DCDCDCDC - multianewArrayEvaluator. More than 2 dimensions, so we can't handle this scenario in JITT'ed code\n");
+      return performHelperCall(node, NULL, TR::acall, false, cg);
+      }
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *nonZeroFirstDimLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *oolFailLabel = generateLabelSymbol(cg);
+
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionStart);
+
+   TR::Register *dimsPtrReg = cg->evaluate(firstChild);
+   TR::Register *dimReg = cg->evaluate(secondChild);
+   TR::Register *classReg = cg->evaluate(thirdChild);
+
+   // First generate the OOL code. This path will be used to handle the scenarios that we don't want to handle in JITT'ed code.
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = new (cg->trHeapMemory()) TR_S390OutOfLineCodeSection(oolFailLabel, cFlowRegionEnd, cg);
+   cg->getS390OutOfLineCodeSectionList().push_front(outlinedSlowPath);
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, oolFailLabel);
+
+   TR::Node *_callNode = createOutlinedCallNode(cg, node, TR::acall);
+   TR::Register *targetReg2 = NULL;
+   targetReg2 = TR::TreeEvaluator::performCall(_callNode, false, cg);
+   cg->decReferenceCount(_callNode);
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, targetReg, targetReg2);
+
+   // After returning from the helper, branch back to mainline code.
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+   // End of OOL
+   outlinedSlowPath->swapInstructionListsWithCompilation();
 
+   // In the mainline, first load the first and second dimensions' lengths into registers.
+   TR::Register *firstDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, firstDimLenReg, generateS390MemoryReference(dimsPtrReg, 4, cg));
+
+   TR::Register *secondDimLenReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::LG, node, secondDimLenReg, generateS390MemoryReference(dimsPtrReg, 0, cg));
+
+
+   // Check to see if second dimension is indeed 0. If yes, then proceed to handle the case here. Otherwise jump to OOL code to perform the call.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, secondDimLenReg, (int8_t)0, oolFailLabel, (TR::InstOpCode::COND_BNE));
+
+   // Now check to see if first dimension is also 0. If no, then jump to nonZeroFirstDimLabel to handle it. Otherwise continue below to handle the case when length for both dimensions is 0.
+   generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, firstDimLenReg, (int8_t)0, nonZeroFirstDimLabel, (TR::InstOpCode::COND_BNE));
```

Comment:
We don't typically generate this instructions by themselves. We should use `generateS390CompareAndBranchInstruction` to mimic all the other uses. Also let's remove the unnecessary casts and extra parenthesis while here.

Reply:
Removing the casts causes an ambiguous call to overloaded function error. Here are the error details:

```
error: call of overloaded ‘generateRIEInstruction(TR::CodeGenerator*&, OMR::InstOpCode::Mnemonic, TR::Node*&, TR::Register*&, int, TR::LabelSymbol*&, OMR::Z::InstOpCode::S390BranchCondition)’ is ambiguous
    generateRIEInstruction(cg, TR::InstOpCode::CLGIJ, node, firstDimLenReg, 0, nonZeroFirstDimLabel, (TR::InstOpCode::COND_BNE));
```
Candidates:
```
S390GenerateInstructions.hpp:913:19: note: candidate: TR::Instruction* generateRIEInstruction(TR::CodeGenerator*, OMR::InstOpCode::Mnemonic, TR::Node*, TR::Register*, TR::Register*, TR::LabelSymbol*, OMR::Z::InstOpCode::S390BranchCondition, TR::Instruction*)
```

```
S390GenerateInstructions.hpp:934:19: note: candidate: TR::Instruction* generateRIEInstruction(TR::CodeGenerator*, OMR::InstOpCode::Mnemonic, TR::Node*, TR::Register*, int8_t, TR::LabelSymbol*, OMR::Z::InstOpCode::S390BranchCondition, TR::Instruction*)
```

Reply:
Right. Special case because of `NULL` and 0 alias. Let's follow an existing example using `generateS390CompareAndBranchInstruction`:
https://github.com/eclipse/openj9/blob/547bb8e04985d8043443b977ef979eb183139da6/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L602

Reply:
This will be fixed in my next commit. We won't need to worry about the casts as I'll be using `generateS390CompareAndBranchInstruction`